### PR TITLE
[nexus-db-queries] make saga_update more resilient, record_event idempotent

### PR DESCRIPTION
**Pull request by sunshowers**
See discussion in https://github.com/oxidecomputer/omicron/issues/2416 and https://github.com/oxidecomputer/omicron/issues/6090#issuecomment-2229509411.

A summary of the changes here:

1. Made `saga_create_event` idempotent. Previously, creating another event that duplicated the one which already existed would fail -- now it succeeds. These events are meant to be an append-only idempotent log, so this is okay. Also added a test for this.
2. `saga_update_state` was already idempotent -- added a test which made sure of this. Also added a comment about how idempotence may not be enough in the future.
3. Added a retry loop around saga state updates, similar to the one around recording events.

### CODE REVIEW

File: nexus/db-queries/src/db/datastore/saga.rs
Comment (hawkw) [inline nexus/db-queries/src/db/datastore/saga.rs:57]: :+1:  this makes sense to me --- I appreciate the comment!

File: nexus/db-queries/src/db/sec_store.rs
Comment (hawkw) [inline nexus/db-queries/src/db/sec_store.rs:155]: nitpicky, unimportant: maybe make the duration a constant?

File: nexus/db-queries/src/db/datastore/saga.rs
Comment (hawkw) [inline nexus/db-queries/src/db/datastore/saga.rs:96]: take it or leave it: this kind of feels like it ought to be API documentation on the method, rather than an internal implementation comment? IIUC it is describing what the caller is expected to do with the function, which feels like API docs to me...

**Review by hawkw — APPROVED** [review]
This looks good to me! I had a couple tiny suggestions, but I don't see any issue with the change overall.

File: nexus/db-queries/src/db/datastore/saga.rs
Comment (sunshowers) [reply nexus/db-queries/src/db/datastore/saga.rs:96]: Seems reasonable.

**Comment by sunshowers** [comment]
Thanks! I'll let one of Sean and Dave also review this just to be sure we've gotten everything we can think of here.

**Comment by hawkw** [comment]
FWIW, it might be worth seeking out a review from someone more qualified with this stuff than I am; I don't actually know this code at all --- I'm not totally sure why  #6090 was assigned to me in the first place, it might've been a mistake?

**Comment by hawkw** [comment]
> Thanks! I'll let one of Sean and Dave also review this just to be sure we've gotten everything we can think of here.

Ah, yup, looks like you replied while I was writing my prior comment. Waiting for a Sean or Dave review seems like a good idea!

**Review by smklein — APPROVED** [review]

**Pull request merged**
